charts/opentelemetry-operator/release/main.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"gopkg.in/yaml.v3"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+const (
+	operatorManifestURL = "https://github.com/open-telemetry/opentelemetry-operator/releases/latest/download/opentelemetry-operator.yaml"
+	collectorCRDPath = "../crds/crd-opentelemetrycollector.yaml"
+)
+
+// K8sObject describes the values a Kubernetes object typically has.
+type K8sObject struct {
+	ApiVersion string `yaml:"apiVersion"`
+	Kind string `yaml:"kind"`
+	Metadata map[interface{}]interface{} `yaml:"metadata"`
+	Spec map[interface{}]interface{} `yaml:"spec"`
+	Status map[interface{}]interface{} `yaml:"status"`
+}
+
+// The main keeps the OpenTelemetry Operator (OTEL Operator) Helm chart configurations up-to-date with the upstream automatically.
+// First, it scrapes the latest OTEL Operator manifest from the OTEL Operator github:
+// https://github.com/open-telemetry/opentelemetry-operator.
+// Then, it will update the OpenTelemetry Collector CRD YAML file, which is the only CRD at this point.
+// At last, it will retrieve the latest image tag from the OTEL Operator manifest and update the values.yaml and Chart.yaml.
+func main() {
+	// Get the OTEL Operator manifest data.
+	resp, err := http.Get(operatorManifestURL)
+	if err != nil {
+		panic(err)
+	}
+
+	// Parse the manifest.
+	dc := yaml.NewDecoder(resp.Body)
+	for {
+		var curObject K8sObject
+		err := dc.Decode(&curObject)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			panic(err)
+		}
+		switch curObject.Kind {
+		case "CustomResourceDefinition":
+			// Update the OTEL Collector CRD YAML file.
+			out, err := yaml.Marshal(curObject)
+			if err != nil {
+				panic(err)
+			}
+			err = os.WriteFile(collectorCRDPath, out, 0644)
+			if err != nil {
+				panic(err)
+			}
+			fmt.Println("The OpenTelemetry Collector CRD update finished. It's up-to-date now.")
+		case "Deployment":
+			// Retrieve the latest image repository and tag of the two container images.
+			managerImage := strings.Split(curObject.Spec["template"].(map[string]interface{})["spec"].
+				(map[string]interface{})["containers"].([]interface{})[0].(map[string]interface{})["image"].(string), ":")
+			kubeRBACProxyImage := strings.Split(curObject.Spec["template"].(map[string]interface{})["spec"].
+				(map[string]interface{})["containers"].([]interface{})[1].(map[string]interface{})["image"].(string), ":")
+			perlCommandStr1 := "s/    repository: quay.io\\/opentelemetry\\/opentelemetry-operator\\n    " +
+				"tag: \\\".*\\\"\\n  resources/    repository: quay.io\\/opentelemetry\\/opentelemetry-operator\\n    tag: \\\"" +
+				managerImage[1] + "\\\"\\n  resources/igs"
+			c1 := exec.Command("perl", "-0777", "-i", "-pe", perlCommandStr1, "../values.yaml")
```

Comment:
Can't you use `regexp` instead? https://pkg.go.dev/regexp

Reply:
The first substitution method came to my mind was perl. So I wrote it this way - a little bit disorderly though caused by the escape character.

Reply:
I'm still concerned about the maintenance and portability of this. This requires Perl to exist and be in the path for people running this command. This might be true for most Unixes, but what about Windows?

How about getting the spec node and parse it using a YAML library? You can then navigate straight to the node you need to change. After that, just reencode it as YAML and override your values.yaml.

Reply:
Windows doesn't have Perl installed by default, but it's easy for maintainers to install the binaries. I will add this Perl requirements to the check-list doc.

> How about getting the spec node and parse it using a YAML library? You can then navigate straight to the node you need to change. After that, just reencode it as YAML and override your values.yaml.

Yes, we can do the work using YAML parse functions and override values.yaml. But the problem is the marshall process will jeopardise the current order of configuration options in values.yaml. It will also make the comments on top of the configurations in a mess. So we decided to use the string substitution approach to update the values.yaml.

Reply:
String substitution is fine, I'm just not sure the proposal is the right/easiest thing. For instance, you could have also a `values.template.yaml`, which will have placeholders. You can then use a regular string replacement with go, writing the outcome to `values.yaml`.

In any case, I'm still uncomfortable with this solution here. If you still think it's the best solution available, I will gladly accept it, especially if someone else chimes in. @dmitryax, you approved this PR before, your confirmation that this is the best solution for the problem would be enough for me.

Reply:
I wrote this auto update logic since you are uncomfortable with Perl string replacement approach. Now this string replacement work will be done by go. Please review it and let me know what you think about this. Thanks!